# ASTROF — Quick Test Notebook (T4 / pipeline verification)



## 0. GPU Check

In [1]:
!nvidia-smi
import torch

print(f"CUDA available: {torch.cuda.is_available()}")
print(f"GPU: {torch.cuda.get_device_name(0) if torch.cuda.is_available() else 'None'}")
print(
    f"VRAM: {torch.cuda.get_device_properties(0).total_memory / 1e9:.1f} GB"
    if torch.cuda.is_available()
    else ""
)

Sat Apr 25 22:11:10 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.211.01             Driver Version: 570.211.01     CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA H200                    Off |   00000000:8D:00.0 Off |                    0 |
| N/A   28C    P0             74W /  700W |       0MiB / 143771MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+-----

CUDA available: True
GPU: NVIDIA H200
VRAM: 150.1 GB


## 1. Install Dependencies

In [2]:
%%capture
# Core training stack
!pip install unsloth trl peft datasets transformers accelerate bitsandbytes

# OpenEnv client
!pip install openenv-core

# Astronomy + API
!pip install astropy openai requests

!pip install "numpy==1.26.4" --force-reinstall -q

## 2. Clone Repository

In [3]:
import os

REPO_URL = "https://github.com/Cosmos-Atom/astrof"
REPO_DIR = "astrof"

if not os.path.exists(REPO_DIR):
    !git clone {REPO_URL}
else:
    !cd {REPO_DIR} && git pull

%cd {REPO_DIR}
!ls

Already up to date.
/teamspace/studios/this_studio/astrof


/home/zeus/miniconda3/envs/cloudspace/lib/python3.12/site-packages/IPython/core/magics/osm.py:417: UserWarning: using dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]


BLOG.md			   client.py	   requirements.txt
Dockerfile		   data		   scripts
EXPLAINER.md		   inference.py    server
POST_TRAINING_STRATEGY.md  models.py	   train_astrof.ipynb
README.md		   openenv.yaml    train_astrof_test.ipynb
__init__.py		   outputs	   unsloth_compiled_cache
__pycache__		   pyproject.toml  uv.lock


## 3. Configuration

Set your HF token to push the trained adapter to the Hub after training.

In [ ]:
import os

# --- Required ---
os.environ["ENV_BASE_URL"] = "https://Cosmosatom-astrof.hf.space"  # live environment
os.environ["MODEL_NAME"] = "unsloth/Qwen3-1.7B-unsloth-bnb-4bit"

# --- Optional: set to push adapter to HF Hub after training ---
# os.environ["HF_TOKEN"]   = "hf_xxx"
# os.environ["HF_REPO_ID"] = "Cosmosatom/astrof-adapter"


import os, sys

sys.path.insert(0, os.getcwd())
sys.path.insert(0, "scripts")


# --- Quick test: only 10 GRPO steps per task ---
import scripts.train_grpo as tg

tg.GRPO_STEPS_PER_TASK = 100
tg.SFT_EPOCHS = 1


## 4. Smoke Test

Verify imports, data files, and environment connectivity before starting training.

In [ ]:
import requests, time

requests.get("https://Cosmosatom-astrof.hf.space/health")  # wake up
time.sleep(3)

In [6]:
!pip install openenv-core astropy -q
!python scripts/smoke_test.py


[1] Import checks
  PASS  novelty_detector imports
  PASS  continual_learner imports
  PASS  models imports
  PASS  client imports

[2] Data checks
         5780 SFT examples, roles: ['coordinator', 'executor', 'planner']
  PASS  sft_warmstart.jsonl exists and has all roles
  PASS  planet CSVs exist

[3] Environment connectivity
         connected to https://Cosmosatom-astrof.hf.space
         planner narrative: SCIENCE PLANNER | Step 0/18 | Telescopes available: 1

TARGETS VISIBLE AND UNOBS...
  PASS  env.reset(easy) connects to https://Cosmosatom-astrof.hf.space

[4] Episode runner (dummy model, 3 steps)
         9 step records, rewards: [0.1, 0.1, 0.1, 0.1, 0.1]
  PASS  _run_episode(easy, dummy_model, 3 steps)

[5] NoveltyDetector end-to-end
         bad step novelty score: 0.250, episode novelty: 0.250
  PASS  NoveltyDetector detects bad step

[6] ContinualLearner dry-run
[CONTINUAL] Novel episode buffered (score=0.90, buffer=2/15)
[CONTINUAL] Novel episode buffered (score=0.90, b

## 5. Load Model (Qwen3-1.7B, 4-bit)

Loads the base model with Unsloth optimizations and attaches a LoRA adapter (r=16).

In [ ]:
import sys, os

sys.path.insert(0, "scripts")

from unsloth import FastLanguageModel

MODEL_NAME = os.environ["MODEL_NAME"]

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name=MODEL_NAME,
    max_seq_length=2048,
    dtype=torch.float16,
    load_in_4bit=True,
)

model = FastLanguageModel.get_peft_model(
    model,
    r=16,
    target_modules=[
        "q_proj",
        "k_proj",
        "v_proj",
        "o_proj",
        "gate_proj",
        "up_proj",
        "down_proj",
    ],
    lora_alpha=16,
    lora_dropout=0,
    bias="none",
    use_gradient_checkpointing=False,
    random_state=2026,
)

print(f"Model loaded: {MODEL_NAME}")
print(
    f"Trainable params: {sum(p.numel() for p in model.parameters() if p.requires_grad):,}"
)

==((====))==  Unsloth 2026.4.8: Fast Qwen3 patching. Transformers: 5.5.0.
   \\   /|    NVIDIA H200. Num GPUs = 1. Max memory: 139.811 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.10.0+cu128. CUDA: 9.0. CUDA Toolkit: 12.8. Triton: 3.6.0
\        /    Bfloat16 = TRUE. FA [Xformers = 0.0.35. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Loading weights:   0%|          | 0/310 [00:00<?, ?it/s]

unsloth/Qwen3-1.7B-unsloth-bnb-4bit does not have a padding token! Will use pad_token = <|PAD_TOKEN|>.
Model loaded: unsloth/Qwen3-1.7B-unsloth-bnb-4bit
Trainable params: 17,432,576


## 6. SFT Warm-start

Fine-tunes one epoch on 5,780 greedy-policy demonstrations (`data/sft_warmstart.jsonl`).
This primes the model to output valid JSON for all 5 role-conditioned prompts before RL begins.
Without this, GRPO sees only zero-reward rollouts in early training.

**Expected time:** ~10 min on A100

In [ ]:
import torch
import train_grpo as tg

# Detect model dtype and patch SFTConfig + GRPOConfig to match — prevents
# Unsloth's fp16/bf16 mismatch error regardless of GPU type.
_model_dtype = next(p for p in model.parameters() if not p.requires_grad).dtype
_use_fp16 = _model_dtype == torch.float16
_use_bf16 = _model_dtype == torch.bfloat16
print(f"Model dtype: {_model_dtype} → fp16={_use_fp16}, bf16={_use_bf16}")

from trl import SFTConfig, GRPOConfig
import functools

# Unwrap any previous patches to avoid recursion
_orig_sft = getattr(SFTConfig.__init__, "__wrapped__", SFTConfig.__init__)
_orig_grpo = getattr(GRPOConfig.__init__, "__wrapped__", GRPOConfig.__init__)


@functools.wraps(_orig_sft)
def _sft_patched(self, *a, **kw):
    kw["bf16"] = _use_bf16
    kw["fp16"] = _use_fp16
    _orig_sft(self, *a, **kw)


@functools.wraps(_orig_grpo)
def _grpo_patched(self, *a, **kw):
    kw["bf16"] = _use_bf16
    kw["fp16"] = _use_fp16
    kw["max_completion_length"] = 512
    _orig_grpo(self, *a, **kw)


SFTConfig.__init__ = _sft_patched
GRPOConfig.__init__ = _grpo_patched
print("Patched SFTConfig + GRPOConfig: dtype-matched, max_completion_length=512")

tg.sft_warmstart(model, tokenizer)
print("SFT warm-start complete.")

Model dtype: torch.float16 → fp16=True, bf16=False
Patched SFTConfig + GRPOConfig: dtype-matched, max_completion_length=512
[SFT] Warm-starting on 5780 examples ...


Unsloth: Tokenizing ["text"] (num_proc=20):   0%|          | 0/5780 [00:00<?, ? examples/s]

The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'bos_token_id': None}.


🦥 Unsloth: Padding-free auto-enabled, enabling faster training.


==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 5,780 | Num Epochs = 1 | Total steps = 362
O^O/ \_/ \    Batch size per device = 4 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (4 x 4 x 1) = 16
 "-____-"     Trainable parameters = 17,432,576 of 1,738,007,552 (1.00% trained)


Step,Training Loss
10,4.273087
20,3.384212
30,2.403058
40,1.741162
50,0.994656
60,0.461574
70,0.228147
80,0.162305
90,0.142628
100,0.122112


[SFT] Warm-start complete.
SFT warm-start complete.


## 7. GRPO Curriculum Training

Trains the model using Group Relative Policy Optimization across 3 difficulty levels:

| Stage | Task | What the agent learns |
|-------|------|-----------------------|
| 1 | easy | Single telescope, JSON format, priority scoring |
| 2 | medium | 3-telescope coordination, duplicate avoidance |
| 3 | hard | Weather recovery, transit deadlines |

**Reward:** `0.9 × science_yield + 0.1 × parse_rate`  
**Expected time:** ~20 min per stage on A100 (100 steps each)

In [ ]:
import os
import sys

sys.path.insert(0, os.getcwd())
os.environ["CUDA_LAUNCH_BLOCKING"] = "1"

import scripts.train_grpo as tg

# patch ENV BEFORE calling train_grpo
tg.ENV_BASE_URL = "https://Cosmosatom-astrof.hf.space"
tg.GRPO_STEPS_PER_TASK = 100
print("ENV:", tg.ENV_BASE_URL)
print("Steps:", tg.GRPO_STEPS_PER_TASK)


for task_id in tg.CURRICULUM:
    print(f"\n{'=' * 50}\nTraining on task: {task_id}\n{'=' * 50}")
    tg.train_grpo(model, tokenizer, task_id)

## 8. Expert Task — ToO + Novel Category

Run this if time allows. Tests Target-of-Opportunity interrupt handling and
in-context adaptation to the `gravitational_wave_host` category.

**Expected time:** ~20 min on A100

In [ ]:
import scripts.train_grpo as tg

# patch ENV BEFORE calling train_grpo
tg.ENV_BASE_URL = "https://Cosmosatom-astrof.hf.space"
tg.GRPO_STEPS_PER_TASK = 100  # quick test
print("ENV:", tg.ENV_BASE_URL)
print("Steps:", tg.GRPO_STEPS_PER_TASK)

tg.train_grpo(model, tokenizer, "expert")
print("Expert task training complete.")

## 9. Continual Learning Loop

Runs 200 post-curriculum episodes on hard+expert tasks.
The `NoveltyDetector` watches every step — when 30 novel episodes accumulate,
`ContinualLearner` automatically trains a new LoRA adapter and merges it into the model.

No human intervention required.

**Expected time:** ~30 min on A100 (may trigger 1-2 adapter merges)

In [ ]:
# # import scripts.train_grpo as tg

# # patch ENV BEFORE calling train_grpo
# tg.ENV_BASE_URL = "https://Cosmosatom-astrof.hf.space"
# tg.GRPO_STEPS_PER_TASK = 20  # quick test
# print("ENV:", tg.ENV_BASE_URL)
# print("Steps:", tg.GRPO_STEPS_PER_TASK)

# tg.GRPO_STEPS_PER_TASK = 20 # quick test

# from train_grpo import run_continual_loop

# model = run_continual_loop(model, tokenizer)
# print("Continual learning loop complete.")

## 10. Save Adapter

In [ ]:
# import os

# os.makedirs("outputs/final", exist_ok=True)
# model.save_pretrained("outputs/final")
# tokenizer.save_pretrained("outputs/final")
# print("Adapter saved to outputs/final/")
# !ls outputs/final/

## 11. Push to HF Hub (Optional)

In [ ]:
# HF_TOKEN = os.environ.get("HF_TOKEN", "")
# HF_REPO_ID = os.environ.get("HF_REPO_ID", "")

# if HF_TOKEN and HF_REPO_ID:
#     model.push_to_hub(HF_REPO_ID, token=HF_TOKEN)
#     tokenizer.push_to_hub(HF_REPO_ID, token=HF_TOKEN)
#     print(f"Pushed to {HF_REPO_ID}")
# else:
#     print("HF_TOKEN or HF_REPO_ID not set — skipping push.")
#     print("Set os.environ['HF_TOKEN'] and os.environ['HF_REPO_ID'] to enable.")

## 12. Benchmark — Before vs After

Run the benchmark script with the trained model to get your AFTER numbers.
Compare against the zero-shot baseline already collected.

**Requires:** Ollama running with the trained model, OR set `API_BASE_URL` to HF Inference.

In [ ]:
# import os
# os.makedirs("outputs/final", exist_ok=True)
# model.save_pretrained("outputs/final")
# tokenizer.save_pretrained("outputs/final")
# print("Saved!")

In [ ]:
# # Option A: run inference directly in this notebook using the trained model
# import sys, os, json, re

# sys.path.insert(0, ".")
# sys.path.insert(0, "scripts")

# from train_grpo import _make_direct_model_fn, _run_episode, MAX_STEPS_PER_TASK

# model_fn = _make_direct_model_fn(model, tokenizer)

# print("Running trained model evaluation...")
# results = {}
# for task in ["easy", "medium", "hard", "expert"]:
#     steps = _run_episode(task, model_fn)
#     rewards = [s["reward"] for s in steps]
#     avg_reward = sum(rewards) / max(len(rewards), 1)
#     results[task] = round(avg_reward, 4)
#     print(f"  {task:8s}  avg_reward={avg_reward:.4f}  steps={len(rewards)}")

# print("\nTrained model results:")
# print(json.dumps(results, indent=2))

## Summary

| Stage | What ran |
|-------|----------|
| SFT warm-start | 5,780 greedy demonstrations, 1 epoch |
| GRPO easy | Single telescope, format + priority learning |
| GRPO medium | 3-telescope coordination, duplicate avoidance |
| GRPO hard | Weather recovery, transit deadlines |
| GRPO expert | ToO interrupts, novel category adaptation |
| Continual loop | 200 episodes, auto LoRA on novelty |

**Adapter saved to:** `outputs/final/`

---
*ASTROF — Team X: Kavya Sree Kammari, Sanyam Bhardwaj, Yasasree Lasya*  
*Meta PyTorch / OpenEnv Hackathon — Grand Finale, April 2026*